### PR TITLE
[firebase-recaptcha] Invisible reCAPTCHA

### PR DESCRIPTION
**Pull request by IjzerenHein**
# Why

This PR adds support for invisible reCAPTCHA. When using invisible reCAPTCHA, the user only shown the reCAPTCHA UI when the system cannot guarantee that the user is not a bot. This allows for a faster sign-up workflow.

ezgif com-gif-maker

### Banner

image

# How

- Add `attemptInvisibleVerification` prop to `<FirebaseRecaptchaVerifierModal>`
- Add `<FirebaseRecaptchaBanner />` component for displaying Google terms
- Add `attemptInvisibleVerification` prop to NCL (native component list)
- Add `<FirebaseRecaptchaBanner />` component to NCL
- Updated docs
- Updated default firebase lib to 8.0.0 (loaded inside web-view, not RN)

# Test Plan

- Start NCL `cd apps/native-component-list && expo start`
- Open `API -> Firebase reCAPTCHA`
- Click on 2nd list-item row to enable `attemptInvisibleVerification` reCAPTCHA
- Click on 1st list-item row to request a reCAPTCHA token
- Token is obtained without showing any challenge UI most of the time
- Verified on iPhone & Android

### CODE REVIEW

**Comment by SimonHausdorf** [comment]
What is missing for this to be done?

**Comment by a-shen** [comment]
> What is missing for this to be done?

+1. I would love to be able to use this functionality too.

**Comment by IjzerenHein** [comment]
Hi! I've added the following comments to the description describing why this is blocked

> Currently this still triggers the full challenge too often. The way it works is that it creates a trusted click event to click a hidden form button. This most likely causes reCAPTCHA to asses that it isn't a bot initially, but will recognise it as a bot after repeated use. More investigation is needed to come up with a strategy that is more reliable.

**Comment by IjzerenHein** [comment]
Whether reCAPTCHA allows for invisible validation is quite unpredictable though. You might validate successfully 5 times in a row, or need to re-validate after each succesful `invisible` validation. This unpredictability reduces the usefulness and might not match the users expectations. @brentvatne what do you think?

**Comment by SimonHausdorf** [comment]
Does expo support silent push notifications? This would avoid the captcha to be shown. I am not sure if an invisible captcha will have any difference to the normal captcha as it might be always shown without having silent push notifications in place.

**Comment by gerardo15** [comment]
Would love to see this getting merged!

**Comment by IjzerenHein** [comment]
> Does expo support silent push notifications? This would avoid the captcha to be shown. I am not sure if an invisible captcha will have any difference to the normal captcha as it might be always shown without having silent push notifications in place.

@SimonHausdorf I think you are referring to the Firebase Auth when using the native SDK. Unfortunately this refers to JavaScript Firebase SDK, and specifically only to the reCAPCHA part of it (which you can also use without phone auth)

**Comment by IjzerenHein** [comment]
@gerardo15 This doesn't seem stable enough and might not meet your expectations. We could release it as experimental and you could try it out in your environment.

File: docs/pages/versions/v39.0.0/sdk/firebase-recaptcha.md [hunk 1]
```diff
@@ -29,11 +29,14 @@ Instead of using the standard `firebase.auth.RecaptchaVerifier` class, we will b
 
 Add the `<FirebaseRecaptchaVerifierModal>` component to your screen and store its ref for later use. Also pass in the Firebase web configuration using the `firebaseConfig` prop.
 
+Optionally you can enable `invisible` reCAPTCHA. When enabled, reCAPTCHA attempts to complete the verification without showing any UI to the user. There is however no guarentee that this will always work. When the user cannot be verified automatically, the full reCATPCHA challenge UI is shown.
+> Google terms apply, use `<FirebaseRecaptchaBanner>` to for instance show the Google terms when using invisible reCAPTCHA.
```
Comment (brentvatne) [inline docs/pages/versions/v39.0.0/sdk/firebase-recaptcha.md:33]:
```suggestion
> Google terms apply - use `<FirebaseRecaptchaBanner>` to show the Google terms when using invisible reCAPTCHA.
```

**Review by brentvatne — COMMENTED** [review]
my suggestion for this would be to split up the invisible recaptcha verifier and mark it as experimental. if it the invisible recaptcha fails, then the developer can then render the original recaptcha verifier modal.

i think this is a more easy understandable api for now: you can opt into the experimental invisible verifier by rendering it, and fallback to the stable but kind of unpleasant modal verifier. the purpose of the props also becomes very obvious - `title` alongside `invisible` is slightly confusing at first glance. we also isolate the experimental code from the stable code by keeping these separate.

if we get to a point where we're confident in the invisible verifier, we could potentially unify into one component and make it so we always attempt invisible verifier before opening the modal (because who wouldn't want that, if it was reliable?) and add a prop like `skipInvisibleVerificationAttempt` as an escape hatch.

**Comment by IjzerenHein** [comment]
> my suggestion for this would be to split up the invisible recaptcha verifier and mark it as experimental. if it the invisible recaptcha fails, then the developer can then render the original recaptcha verifier modal.
> 
> i think this is a more easy understandable api for now: you can opt into the experimental invisible verifier by rendering it, and fallback to the stable but kind of unpleasant modal verifier. the purpose of the props also becomes very obvious - `title` alongside `invisible` is slightly confusing at first glance. we also isolate the experimental code from the stable code by keeping these separate.
> 
> if we get to a point where we're confident in the invisible verifier, we could potentially unify into one component and make it so we always attempt invisible verifier before opening the modal (because who wouldn't want that, if it was reliable?) and add a prop like `skipInvisibleVerificationAttempt` as an escape hatch.

Thanks for the feedback Brent 👍 

I've given this some thought and have renamed `invisible` to `attemptInvisibleVerification`. This prop is marked as *experimental* in the documentation. This should remove the confusion with `title` and the fact that it doesn't make everything invisible.

Decided not to include a separate "experimental" component. Doing so would also require users to update their code once we remove/rename it, which is something we should try to avoid. And having users switch to visible reCAPTCHA manually once invisible fails, seems like a workflow that's easy to mess up and would require us to expose more internal APIs.

File: packages/expo-firebase-recaptcha/src/FirebaseRecaptchaVerifierModal.tsx [hunk 2]
```diff
@@ -77,9 +113,12 @@ export default class FirebaseRecaptchaVerifierModal extends React.Component<Prop
     if (resolve) {
       resolve(token);
     }
-    this.setState({
+    this.setState(state => ({
```
[the remark this answers is not in the document]
Comment (IjzerenHein) [reply packages/expo-firebase-recaptcha/src/FirebaseRecaptchaVerifierModal.tsx:120]: it's not a linear state, sometimes flags are better for these cases

**Pull request merged**